File: packages/worker/src/services/image-fetcher.ts
/**
 * Image Fetcher Orchestrator
 *
 * Reads [IMAGE: keyword] entries from scenes.json, routes requests to
 * Pexels (photos) or Freepik (illustrations), downloads images to the
 * Remotion workspace, and updates scene data with local file paths.
 *
 * Pipeline: Director → scenes.json → Image Fetcher → updated scenes.json + images → Animator
 */

import { mkdir } from 'fs/promises';
import { join, extname } from 'path';
import { logger } from '../logger.js';
import { searchPhotos, downloadPhoto, type PexelsPhoto } from './pexels.js';
import { searchResources, downloadResource, type ResourceOption } from './freepik.js';

// =============================================================================
// Types
// =============================================================================

/** Image entry as written by the Director in scenes.json */
export interface SceneImageRequest {
  keyword: string;
  type: 'photo' | 'illustration';
  purpose: 'hero' | 'accent' | 'background';
  description?: string;
  placement?: 'center' | 'background' | 'left' | 'right';
}

/** Image entry after fetching — includes local paths and attribution */
export interface FetchedImage extends SceneImageRequest {
  localPath: string;
  remotionPath: string;
  source: 'pexels' | 'freepik';
  attribution: string;
  width?: number;
  height?: number;
}

/** A scene object from scenes.json (partial — only fields we care about) */
interface SceneEntry {
  id: number | string;
  name?: string;
  images?: SceneImageRequest[];
  [key: string]: unknown;
}

/** The top-level scenes.json structure */
interface ScenesData {
  scenes: SceneEntry[];
  [key: string]: unknown;
}

/** Image option for plan pipeline (before download — used in plan-visuals.ts) */
export interface ImageOption {
  id: string;
  title: string;
  thumbnailUrl: string;
  source: 'pexels' | 'freepik';
  type: 'photo' | 'illustration';
}

// =============================================================================
// Constants
// =============================================================================

const MAX_IMAGES_PER_SCENE = 2;
const MAX_TOTAL_IMAGES = 10;
const CONCURRENCY_LIMIT = 5;
const TOTAL_TIMEOUT_MS = 60_000;

// =============================================================================
// Main Orchestrator
// =============================================================================

/**
 * Fetch images for all scenes and download them to the workspace.
 *
 * @param scenesData - Parsed scenes.json content
 * @param workspacePath - Root workspace path (images go into public/assets/images/)
 * @returns Updated scenesData with localPath/remotionPath populated on fetched images
 */
export async function fetchImagesForScenes(
  scenesData: ScenesData,
  workspacePath: string,
): Promise<{ updatedScenes: ScenesData; downloadedCount: number }> {
  const imagesDir = join(workspacePath, 'public', 'assets', 'images');
  await mkdir(imagesDir, { recursive: true });

  // Collect all image requests across scenes (respecting budgets)
  const tasks: Array<{
    sceneIndex: number;
    imageIndex: number;
    request: SceneImageRequest;
  }> = [];

  for (let si = 0; si < scenesData.scenes.length; si++) {
    const scene = scenesData.scenes[si];
    if (!scene.images || !Array.isArray(scene.images)) continue;

    // Cap per-scene
    const sceneImages = scene.images.slice(0, MAX_IMAGES_PER_SCENE);
    for (let ii = 0; ii < sceneImages.length; ii++) {
      if (tasks.length >= MAX_TOTAL_IMAGES) break;
      tasks.push({ sceneIndex: si, imageIndex: ii, request: sceneImages[ii] });
    }
    if (tasks.length >= MAX_TOTAL_IMAGES) break;
  }

  if (tasks.length === 0) {
    logger.info('No image requests found in scenes — skipping image fetch');
    return { updatedScenes: scenesData, downloadedCount: 0 };
  }

  logger.info({ taskCount: tasks.length }, 'Fetching images for scenes');

  // Execute with concurrency limit and total timeout
  let downloadedCount = 0;

  const results = await Promise.race([
    runWithConcurrency(
      tasks,
      async (task) => {
        const sceneId = scenesData.scenes[task.sceneIndex].id ?? task.sceneIndex + 1;
        const slug = slugify(task.request.keyword);
        const fetched = await fetchSingleImage(
          task.request,
          imagesDir,
          `scene${sceneId}-${task.request.purpose}-${slug}`,
        );
        return { ...task, fetched };
      },
      CONCURRENCY_LIMIT,
    ),
    timeoutPromise(TOTAL_TIMEOUT_MS),
  ]);

  // Apply results back to scenesData
  if (Array.isArray(results)) {
    for (const result of results) {
      if (result.fetched) {
        const scene = scenesData.scenes[result.sceneIndex];
        if (!scene.images) scene.images = [];
        // Replace the request entry with the fetched version
        scene.images[result.imageIndex] = result.fetched;
        downloadedCount++;
      }
    }
  }

  // Clean up any image entries that weren't fetched (remove entries without remotionPath)
  for (const scene of scenesData.scenes) {
    if (scene.images && Array.isArray(scene.images)) {
      scene.images = scene.images.filter(
        (img) => 'remotionPath' in img && (img as FetchedImage).remotionPath,
      ) as SceneImageRequest[];
    }
  }

  logger.info({ downloadedCount, total: tasks.length }, 'Image fetching complete');
  return { updatedScenes: scenesData, downloadedCount };
}

// =============================================================================
// Plan Pipeline — Fetch image options (thumbnails only, no download)
// =============================================================================

/**
 * For the plan pipeline (plan-visuals.ts): search for image options per scene
 * and attach thumbnails for user preview. Does NOT download full images.
 */
export async function fetchImageOptionsForPlan<T extends { scenes?: unknown; [key: string]: unknown }>(
  planData: T,
): Promise<T & { imageOptions?: Record<string, Record<string, ImageOption[]>> }> {
  const scenesObj = planData.scenes as Record<string, unknown> | undefined;
  const scenesArray = ((scenesObj as any)?.scenes as SceneEntry[]) || [];

  // Collect unique keywords mapped to scene IDs and types
  const keywordMap = new Map<string, { sceneIds: string[]; type: 'photo' | 'illustration' }>();

  for (const scene of scenesArray) {
    if (!scene.images || !Array.isArray(scene.images)) continue;
    const sceneId = String(scene.id ?? scene.name ?? '');
    for (const img of scene.images) {
      const key = `${img.keyword}:${img.type}`;
      const existing = keywordMap.get(key);
      if (existing) {
        existing.sceneIds.push(sceneId);
      } else {
        keywordMap.set(key, { sceneIds: [sceneId], type: img.type });
      }
    }
  }

  if (keywordMap.size === 0) {
    logger.info('No image keywords found in plan — skipping image options fetch');
    return planData;
  }

  logger.info({ keywordCount: keywordMap.size }, 'Fetching image options for plan');

  // Fetch options for each unique keyword
  const entries = [...keywordMap.entries()];
  const results = await Promise.all(
    entries.map(async ([key, { sceneIds, type }]) => {
      const keyword = key.split(':')[0];
      const options = await searchImageOptions(keyword, type, 3);
      return { keyword, sceneIds, options };
    }),
  );

  // Build imageOptions map: { sceneId: { keyword: ImageOption[] } }
  const imageOptions: Record<string, Record<string, ImageOption[]>> = {};
  for (const { keyword, sceneIds, options } of results) {
    if (options.length === 0) continue;
    for (const sceneId of sceneIds) {
      if (!imageOptions[sceneId]) imageOptions[sceneId] = {};
      imageOptions[sceneId][keyword] = options;
    }
  }

  logger.info({ sceneCount: Object.keys(imageOptions).length }, 'Image options fetched for plan');
  return { ...planData, imageOptions };
}

// =============================================================================
// Internal Helpers
// =============================================================================

async function fetchSingleImage(
  request: SceneImageRequest,
  imagesDir: string,
  filePrefix: string,
): Promise<FetchedImage | null> {
  try {
    if (request.type === 'photo') {
      return await fetchPexelsPhoto(request, imagesDir, filePrefix);
    } else {
      return await fetchFreepikIllustration(request, imagesDir, filePrefix);
    }
  } catch (err) {
    logger.warn({ err, keyword: request.keyword, type: request.type }, 'Image fetch failed — skipping');
    return null;
  }
}

async function fetchPexelsPhoto(
  request: SceneImageRequest,
  imagesDir: string,
  filePrefix: string,
): Promise<FetchedImage | null> {
  const photos = await searchPhotos(request.keyword, { perPage: 3 });
  if (photos.length === 0) return null;

  // Pick the first result
  const photo = photos[0];
  const filename = `${filePrefix}.jpg`;
  const destPath = join(imagesDir, filename);

  const dims = await downloadPhoto(photo, destPath);
  if (!dims) return null;

  return {
    ...request,
    localPath: destPath,
    remotionPath: `assets/images/${filename}`,
    source: 'pexels',
    attribution: `Photo by ${photo.photographer} on Pexels`,
    width: dims.width,
    height: dims.height,
  };
}

async function fetchFreepikIllustration(
  request: SceneImageRequest,
  imagesDir: string,
  filePrefix: string,
): Promise<FetchedImage | null> {
  const resources = await searchResources(request.keyword, 3, 'vector');
  if (resources.length === 0) return null;

  const resource = resources[0];
  const filename = `${filePrefix}.jpg`;
  const destPath = join(imagesDir, filename);

  const success = await downloadResource(resource.id, destPath);
  if (!success) return null;

  return {
    ...request,
    localPath: destPath,
    remotionPath: `assets/images/${filename}`,
    source: 'freepik',
    attribution: `Illustration from Freepik`,
  };
}

async function searchImageOptions(
  keyword: string,
  type: 'photo' | 'illustration',
  count: number,
): Promise<ImageOption[]> {
  if (type === 'photo') {
    const photos = await searchPhotos(keyword, { perPage: count });
    return photos.map((p) => ({
      id: String(p.id),
      title: p.alt || keyword,
      thumbnailUrl: p.src.small,
      source: 'pexels' as const,
      type: 'photo' as const,
    }));
  } else {
    const resources = await searchResources(keyword, count, 'vector');
    return resources.map((r) => ({
      id: r.id,
      title: r.title,
      thumbnailUrl: r.thumbnailUrl,
      source: 'freepik' as const,
      type: 'illustration' as const,
    }));
  }
}

function slugify(text: string): string {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, '-')
    .replace(/^-|-$/g, '')
    .slice(0, 30);
}

/**
 * Run async tasks with a concurrency limit.
 */
async function runWithConcurrency<T, R>(
  items: T[],
  fn: (item: T) => Promise<R>,
  limit: number,
): Promise<R[]> {
  const results: R[] = [];
  const executing = new Set<Promise<void>>();

  for (const item of items) {
    const p = fn(item).then((r) => {
      results.push(r);
    });
    const tracked = p.then(() => { executing.delete(tracked); });
    executing.add(tracked);

    if (executing.size >= limit) {
      await Promise.race(executing);
    }
  }

  await Promise.all(executing);
  return results;
}

function timeoutPromise(ms: number): Promise<null> {
  return new Promise((resolve) => {
    setTimeout(() => {
      logger.warn({ timeoutMs: ms }, 'Image fetch total timeout reached');
      resolve(null);
    }, ms);
  });
}

File: packages/worker/src/agents/visual_generator/_image_fetcher.py
"""Image fetching mixin — downloads scene images from Pexels/Freepik."""

import io
import json
import os
import re
import zipfile


class ImageFetcherMixin:
    """Mixin providing _fetch_scene_images for ClaudeVisualGenerator."""

    async def _fetch_scene_images(self) -> int:
        """
        Fetch images for scenes based on the Director's [IMAGE: keyword] entries.

        Reads scenes.json, downloads photos from Pexels and illustrations from Freepik,
        saves them to public/assets/images/, and updates scenes.json in-place.

        Returns the count of successfully downloaded images.
        """
        import httpx

        scenes_json_path = self.src_dir / "scenes.json"
        if not scenes_json_path.exists():
            print("[ClaudeGenerator] No scenes.json found — skipping image fetch")
            return 0

        with open(scenes_json_path, encoding="utf-8") as f:
            scenes_data = json.load(f)

        scenes = scenes_data.get("scenes", [])
        if not scenes:
            return 0

        # Collect all image requests
        image_tasks = []
        for si, scene in enumerate(scenes):
            images = scene.get("images", [])
            if not isinstance(images, list):
                continue
            for ii, img in enumerate(images[:2]):  # Max 2 per scene
                if len(image_tasks) >= 10:  # Max 10 total
                    break
                keyword = img.get("keyword", "")
                img_type = img.get("type", "photo")
                purpose = img.get("purpose", "accent")
                if keyword:
                    image_tasks.append({
                        "scene_index": si,
                        "image_index": ii,
                        "keyword": keyword,
                        "type": img_type,
                        "purpose": purpose,
                    })
            if len(image_tasks) >= 10:
                break

        if not image_tasks:
            print("[ClaudeGenerator] No image requests in scenes — skipping")
            return 0

        print(f"[ClaudeGenerator] Fetching {len(image_tasks)} images for scenes...")

        # Create images directory
        images_dir = self.workspace / "public" / "assets" / "images"
        images_dir.mkdir(parents=True, exist_ok=True)

        pexels_api_key = os.environ.get("PEXELS_API_KEY", "")
        freepik_api_key = os.environ.get("FREEPIK_API_KEY", "")

        downloaded = 0

        async with httpx.AsyncClient(timeout=30.0) as client:
            for task in image_tasks:
                try:
                    scene_id = scenes[task["scene_index"]].get("id", task["scene_index"] + 1)
                    slug = re.sub(r'[^a-z0-9]+', '-', task["keyword"].lower()).strip('-')[:30]
                    filename = f"scene{scene_id}-{task['purpose']}-{slug}.jpg"
                    dest_path = images_dir / filename

                    if task["type"] == "photo" and pexels_api_key:
                        downloaded += await self._fetch_pexels_photo(
                            client, task, scenes, dest_path, filename, pexels_api_key
                        )

                    elif task["type"] == "illustration" and freepik_api_key:
                        downloaded += await self._fetch_freepik_illustration(
                            client, task, scenes, dest_path, filename, freepik_api_key
                        )

                except Exception as e:
                    print(f"[ClaudeGenerator] Image fetch failed for '{task['keyword']}': {e}")
                    continue

        # Remove image entries that weren't successfully fetched
        for scene in scenes:
            if "images" in scene and isinstance(scene["images"], list):
                scene["images"] = [
                    img for img in scene["images"]
                    if isinstance(img, dict) and img.get("remotionPath")
                ]

        # Write updated scenes.json
        with open(scenes_json_path, "w", encoding="utf-8") as f:
            json.dump(scenes_data, f, indent=2)

        print(f"[ClaudeGenerator] Image fetch complete: {downloaded}/{len(image_tasks)} downloaded")
        return downloaded

    async def _fetch_pexels_photo(self, client, task, scenes, dest_path, filename, api_key) -> int:
        """Fetch a photo from Pexels. Returns 1 on success, 0 on failure."""
        resp = await client.get(
            "https://api.pexels.com/v1/search",
            params={"query": task["keyword"], "per_page": "3"},
            headers={"Authorization": api_key},
        )
        if resp.status_code != 200:
            return 0
        data = resp.json()
        photos = data.get("photos", [])
        if not photos:
            return 0

        photo = photos[0]
        photo_url = photo.get("src", {}).get("large", "")
        if not photo_url:
            return 0

        dl_resp = await client.get(photo_url)
        if dl_resp.status_code != 200:
            return 0
        dest_path.write_bytes(dl_resp.content)

        img_entry = scenes[task["scene_index"]]["images"][task["image_index"]]
        img_entry["localPath"] = str(dest_path)
        img_entry["remotionPath"] = f"assets/images/{filename}"
        img_entry["source"] = "pexels"
        img_entry["attribution"] = f"Photo by {photo.get('photographer', 'Unknown')} on Pexels"
        img_entry["width"] = photo.get("width")
        img_entry["height"] = photo.get("height")
        print(f"[ClaudeGenerator] Downloaded photo: {filename}")
        return 1

    async def _fetch_freepik_illustration(self, client, task, scenes, dest_path, filename, api_key) -> int:
        """Fetch an illustration from Freepik. Returns 1 on success, 0 on failure."""
        resp = await client.get(
            "https://api.freepik.com/v1/resources",
            params={
                "term": task["keyword"],
                "limit": "3",
                "filters[content_type][vector]": "1",
            },
            headers={
                "x-freepik-api-key": api_key,
                "Accept": "application/json",
            },
        )
        if resp.status_code != 200:
            return 0
        data = resp.json()
        resources = data.get("data", [])
        if not resources:
            return 0

        resource = resources[0]
        resource_id = str(resource.get("id", ""))
        if not resource_id:
            return 0

        dl_info_resp = await client.get(
            f"https://api.freepik.com/v1/resources/{resource_id}/download",
            headers={
                "x-freepik-api-key": api_key,
                "Accept": "application/json",
            },
        )
        if dl_info_resp.status_code != 200:
            return 0
        dl_info = dl_info_resp.json()
        dl_url = dl_info.get("data", {}).get("url", "")
        if not dl_url:
            return 0

        dl_resp = await client.get(dl_url)
        if dl_resp.status_code != 200:
            return 0

        raw_bytes = dl_resp.content

        # Freepik returns ZIP archives containing image + vector sources.
        # Extract the largest raster image from the ZIP.
        if len(raw_bytes) >= 4 and raw_bytes[:4] == b'PK\x03\x04':
            image_exts = {'.jpg', '.jpeg', '.png', '.webp', '.gif'}
            try:
                with zipfile.ZipFile(io.BytesIO(raw_bytes)) as zf:
                    image_entries = [
                        info for info in zf.infolist()
                        if not info.is_dir()
                        and os.path.splitext(info.filename)[1].lower() in image_exts
                        and info.file_size > 0
                    ]
                    image_entries.sort(key=lambda e: e.file_size, reverse=True)
                    if image_entries:
                        extracted = zf.read(image_entries[0].filename)
                        print(f"[ClaudeGenerator] Extracted {image_entries[0].filename} ({len(extracted)} bytes) from Freepik ZIP")
                        raw_bytes = extracted
                    else:
                        print(f"[ClaudeGenerator] Freepik ZIP has no raster images, skipping: {[i.filename for i in zf.infolist()]}")
                        return 0
            except zipfile.BadZipFile:
                print(f"[ClaudeGenerator] Bad ZIP from Freepik, saving raw download")

        dest_path.write_bytes(raw_bytes)

        img_entry = scenes[task["scene_index"]]["images"][task["image_index"]]
        img_entry["localPath"] = str(dest_path)
        img_entry["remotionPath"] = f"assets/images/{filename}"
        img_entry["source"] = "freepik"
        img_entry["attribution"] = "Illustration from Freepik"
        print(f"[ClaudeGenerator] Downloaded illustration: {filename}")
        return 1
